CLI to interact with Merkle Tree

JereSalo opened the pull request:
Implementation of CLI. 
Run the program with cargo run and then interact with the CLI using the tree command. Valid operations are shown in display. See the `readme` file for more information.

JereSalo committed:
add basic cli structure

JereSalo committed:
add build functionality CLI

JereSalo committed:
add other functionalities to CLI

JereSalo committed:
add verify for CLI

JereSalo committed:
remove and order some comments

JereSalo committed:
make improvements in error handling

JereSalo committed:
fix clippy issues and run cargo fmt

JereSalo committed:
move cli to another module

JereSalo committed:
refactor cli

JereSalo committed:
run cargo clippy and cargo fmt

JereSalo committed:
add readme

JereSalo committed:
update readme

iinaki reviewed (changes requested):
While I was testing the CLI I noticed there is no option to send an already hashed element, you could add a flag to the `add` command, something like `-h`, to indicate that the element doesn't need to be hashed. By doing this you could take more advantage of the `build_without_hashing` fn.

iinaki reviewed (changes requested):
Other thing I noticed is that the tree allows the user to insert multiple times the same element, I think this shouldn't be allowed.

iinaki reviewed (changes requested):
I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? 
I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way.
Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have.
Let me know what you think about this one.

iinaki commented on src/cli.rs:
```diff
@@ -0,0 +1,153 @@
+        loop {
+            // Display prompt
+            print!("tree> ");
+            io::stdout().flush()?; // Flush prompt to the terminal
```
For example here what I would do is the following:

```suggestion
            io::stdout().flush().map_err(|e| MerkleError::FailedToFlush(e.to_string()))?; // Flush prompt to the terminal
```

iinaki commented on src/cli.rs:
```diff
@@ -0,0 +1,153 @@
+        Ok(())
+    }
+
+    fn execute_command(&mut self, command: Commands) -> Result<()> {
```
Here I would do this:

```suggestion
    fn execute_command(&mut self, command: Commands) -> Result<(), MerkleError> {
```

JereSalo replied:
I agree. Done!

iinaki reviewed (changes requested):
Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using. 

When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type. 

By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.

JereSalo committed:
allow adding hashed elements to tree

JereSalo committed:
prevent users from adding repeated elements

JereSalo committed:
add check for duplicates in build

JereSalo committed:
refactor build method

JereSalo committed:
change CLI build, now option -H available

JereSalo committed:
add side enum

JereSalo committed:
add context to errors

JereSalo committed:
add custom error and remove anyhow crate

JereSalo committed:
organize code

JereSalo committed:
make it more beautiful

JereSalo committed:
run cargo fmt

JereSalo commented:
> While I was testing the CLI I noticed there is no option to send an already hashed element, you could add a flag to the `add` command, something like `-h`, to indicate that the element doesn't need to be hashed. By doing this you could take more advantage of the `build_without_hashing` fn.

Done! I implemented it both for `build` and `add`, but using `-H` instead of `-h` because the latter is for help. 
Also I unified the build functions into one because I like to have those in the same place. Now it is just `build` and it receives a bool alongside the elements or hashes. The only problem that I see with what I did is that now I can't say the argument is either a hash or an element, because it can be both. But despite that, I like how it looks.

JereSalo commented:
> Other thing I noticed is that the tree allows the user to insert multiple times the same element, I think this shouldn't be allowed.

Fixed! Both in `build` and `add`.

JereSalo commented:
> I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way. Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have. Let me know what you think about this one.

Actually the CLI does use the `verify` function in the Merkle Tree. You send the `hash` you want to verify and it's `proof`. The CLI parses the proof into a `Vec<ProofElem>` and then calls the `verify` method in Merkle Tree with the given hash and the now parsed proof.

JereSalo commented:
> Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using.
> 
> When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type.
> 
> By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.

I understand your point of view. I changed the way I handle errors and removed the anyhow crate but I didn't map IO errors to be MerkleErrors. I prefer to keep them as IO errors and leave the MerkleErrors just for the errors related to the Merkle Tree specifically. I hope we find a middle ground, but do tell me if you think I'm completely wrong.

JereSalo committed:
update readme and change comments

JereSalo committed:
Delete .DS_Store

iinaki commented on src/proof_element.rs:
```diff
@@ -1,13 +1,19 @@
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct ProofElement {
     pub hash: String,
```
Here hash can be of the Hash type right?

JereSalo replied:
Done!

iinaki commented on src/cli.rs:
```diff
@@ -0,0 +1,170 @@
+use crate::{
+    merkle_error::MerkleError, merkle_tree::MerkleTree, proof_element::ProofElement, side::Side,
+};
+// use anyhow::{Context, Error, Result};
```
You left a comment here!

JereSalo replied:
Good catch!

iinaki commented on README.md:
````diff
@@ -1 +1,21 @@
+```tree build <elements> [--hashed]``` – Builds the tree with the given elements. Use the ```--hashed``` flag if the elements are already hashed.  
+```tree add <element> [--hashed]``` – Adds an element to the tree. Use the ```--hashed``` flag if the element is already hashed.  
+```tree show``` – Displays the tree structure on screen.  
+```tree verify <hash> <proof_file>``` – Verifies a if a proof is correct for a given hash in the tree.  
````
```suggestion
```tree verify <hash> <proof_file>``` – Verifies if a proof is correct for a given hash in the tree.  
```

JereSalo replied:
Done!

iinaki commented:
> > I don't understand much the `verify` function of the CLI, the point is to have a proof in a txt file and then use that file to check if the proof is valid? I also notice that the `verify` function of the `MerkleTree` is not being used, only in the tests. I believe that the `verify` fn of the CLI should call this one, it is much clearer this way. Then you could have another functionality in the CLI to verify if a proof is valid, something like `verify-proof` that uses the stuff that you already have. Let me know what you think about this one.
> 
> Actually the CLI does use the `verify` function in the Merkle Tree. You send the `hash` you want to verify and it's `proof`. The CLI parses the proof into a `Vec<ProofElem>` and then calls the `verify` method in Merkle Tree with the given hash and the now parsed proof.

You're right I missed it, looks perfect to me!

iinaki commented:
> > Regarding the handling of errors, I think you did a great job in the `MerkleTree`, but now it seems a bit odd. I think you should stick to the `-> Result<(), MerkleError>` style that you where using.
> > When I am in a function where I would like to return a custom error enum and I encounter a function that returns its own error type, for example `io::stdout().flush()` what I do is to map the error into my own custom error type.
> > By applying this I believe the error handling looks much cleaner. Also I am not familiar with the `anyhow` crate, I don't know what it does, but you could get rid of it.
> 
> I understand your point of view. I changed the way I handle errors and removed the anyhow crate but I didn't map IO errors to be MerkleErrors. I prefer to keep them as IO errors and leave the MerkleErrors just for the errors related to the Merkle Tree specifically. I hope we find a middle ground, but do tell me if you think I'm completely wrong.

No, it looks very logical to me now! It's fine to separate them in this case actually because the two error kinds mean completely different things and should be managed differently. It wouldn't make sense to end the execution due to a command error from the user, but it does if there was an error with the i/o handling of the stdin, and this is what you are doing!

iinaki reviewed (approved):
Left a few comments, just typos, but other than that great work! I like how you handled the parsing of the commands in the CLI with `try_parse_from`! 🚀🚀🚀

JereSalo committed:
make subtle changes
